### ISSUE TITLE
Extend constant pools to support f32/f64 too

### ISSUE THREAD

**Issue by bnjbvr** [bytecodealliance/wasmtime#1116]
As of today, there's a concept of constant pools in Cranelift, that works for SIMD 128-bits values. I'd like to extend it to work also for float32 and float64 values, to avoid the need for the bitcast, which I'm sure is pretty terrible performance-wise (and doesn't help register allocation either, since it requires one GPR register with a very short lifetime).

My current plan is the following:

- the `uimm128` immediate could be reused to store any kind of constants, since it contains an `ir::Constant` (handle to a vector of bytes of any size); it would be renamed to `constant`.
- a new instruction `load_from_constant_pool` would be introduced, taking an input `constant` value.
- vconst right now can be inserted in the constant pool only from the parser. It'd be nice if this were done during codegen itself. We could recycle "expand_fconst" to do this and legalize vconst / f32const / f64const into a `load_from_constant_pool` with the right Constant.
- we switch the encodings of vconst to work for load_from_constant_pool on the same types; and we add encodings for f32/f64.

Aligning the constant pool values would be nice to have but not required at this point; it would just be a performance fault on x86.

@abrown Does it sound reasonable to you?

**Comment by bjorn3** [comment]
The new backend framework supports this, right?

**Comment by cfallin** [comment]
Yes, the new x64 backend supports constant pools, and uses them at least for 128-bit vector values. It looks like right now we load literal values into integer registers and then move to XMM registers to load 32/64-bit floats, but we could pretty easily change that (using the `XmmLoadConst` VCode instruction) if we wanted to.

I think we can leave this open to track an evaluation of whether this makes sense to do, though I suspect it's not too important for most code's performance; the exception would be if we have a lot of duplicated constants and dedup'ing gives a meaningful reduction in code size / icache misses.

**Comment by fitzgen** [comment]
I don't think this is an issue anymore, the constant pool supports arbitrary constant data. If there are particular lowerings we want to discuss, we can file more-targeted issues for them.
